### PR TITLE
chore(opentelemetry): add initial support for the open telemetry trace api

### PR DESCRIPTION
**Pull request by mabdinur**
# Description

Reviewing this PR might look daunting but it's 80-90% snapshot tests and in code documentation. This PR introduces the minimum required support for the opentelemetry-api. The components introduced by this PR are interdependent and should not be merged separately.

## Overview

This PR introduces components which implement the [opentelemetry-api](https://opentelemetry.io/docs/instrumentation/python/). These components will allow dd-trace-py users to use Open Telemetry standards and libraries to generate and submit datadog traces.

Note - This PR should NOT update the dd-trace-py public api and should NOT impact on existing products/features. Open Telemetry support is disabled by default (for now) and all code artifacts are introduced in an internal module.  OpenTelemetry support will be publicly announced and documented in a future PR.

## Core Components

This PR introduces the following classes which implement abstract classes defined in the Open Telemetry API:  Span, Tracer, TracerProvider, and a RuntimeContext provider. 

### Span

`ddtrace._opentelemetry.span.Span` implements all methods defined in [opentelemetry.trace.Span](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/trace/span.py#L57). This class provides an OpenTelemetry shim for Datadog Spans. All Open Telemetry span operations are executed on an underlying Datadog Span. Span links and Span events are not yet supported by the Datadog API and are therefore out of the scope of this PR.

### Tracer

`ddtrace._opentelemetry.tracer.Tracer` implements the `start_span` and `start_as_current_span` methods defined in [opentelemetry.trace.Tracer](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/trace/__init__.py#L269). `start_span` and `start_as_current_span` return an instance of the Open Telemetry compatible Span described above. All instances of this class uses the [global Datadog Tracer](https://github.com/DataDog/dd-trace-py/blob/v1.8.0/ddtrace/__init__.py#L24) to create and activate spans.

### TracerProvider

`ddtrace._opentelemetry.tracer.TracerProvider` implements the `get_tracer` method defined in [opentelemetry.trace.TracerProvider](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/trace/__init__.py#L188). `get_tracer()` will return an instance of the Open Telemetry compatible tracer described above.  One TracerProvider is configured per application. The TracerProvider should be set on application start up using [opentelemetry.tracer.set_tracer_provider()](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/trace/__init__.py#L524). For ddtrace-run users this is handled by default. 

### DDRuntimeContext

`ddtrace._opentelemetry._context.DDRuntimeContext` implements the Open Telemetry [runtime context interface](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/context/context.py#L24). This component allows the Open Telemetry API and the Datadog API to be used interchangeably. A span activated in one api can be accessed by the other (ex: [ddtrace.tracer.current_span()](https://github.com/DataDog/dd-trace-py/blob/v1.8.0/ddtrace/tracer.py#L839) and [openelemetry.context.get_current()](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/context/__init__.py#L129) will always reference the same Datadog Span or Datadog Context object. 

When an Open Telemetry Span is initialized `DDRuntimeContext` retrieves the active Datadog Span/Context and ensures the newly created Open Telemetry Span has the correct parenting. When an Open Telemetry Span is activated the corresponding Datadog Span or Datadog Context object is stored in the [active ddtrace Context Provider](https://github.com/DataDog/dd-trace-py/blob/v1.8.0/ddtrace/tracer.py#L202). In all cases, the implementation and performance of Datadog [Context Providers](https://github.com/DataDog/dd-trace-py/blob/v1.8.0/ddtrace/provider.py#L24) should NOT be impacted by the Open Telemetry API. In a way, spans generated by the Datadog API are treated as "first class citizens" while spans generated by the ddtrace implementation of Open Telemetry API have the additional overhead of being converted to/from Datadog Spans. This reduces the impact of this feature on existing products and isolates all implementation details to the `ddtrace._opentelemetry` package.

DDRuntimeContext replaces the default context provider ([ContextVarsRuntimeContext](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/context/contextvars_context.py#L19)) provided by the opentelemetry-api via  [an entry point](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/context/__init__.py#L62) and setting the [OTEL_PYTHON_CONTEXT](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/context/__init__.py#L55) environment variable. This environment variable is set in `ddtrace/sitecustomize/bootstrap.py` and is  set when `ddtrace-run` is used (manual instrumentation users will need to set this environment variable in their application, this will documented in a future PR).

### Snapshot (Integration) tests

- For all Tracer and Span operations
- Generating traces using the Datadog and Open Telemetry API interchangeably (tests DDRuntimeContext)
- Generating traces in gevent and asyncio coroutines

### Unit tests

- Configuring the Datadog Open Telemetry TraceProvider.

### Local testing

- Traced a sample app using the OpenTelemetry API with the Datadog TracerProvider and Runtime Context Provider configured.  

## Sample Usage

- Install ddtrace and ensure the datadog agent is installed and running: https://docs.datadoghq.com/agent/
- Run your application with the following command: `ddtrace-run <my_program>`. The `ddtrace-run` command will initialize and set the ddtrace Opentelemetry [TraceProvider](https://github.com/DataDog/dd-trace-py/blob/602c775373a837fc4a4afed7413884b771a0f180/ddtrace/bootstrap/sitecustomize.py#L125) and the `OTEL_PYTHON_CONTEXT` environment variable to supported values. 
- Datadog Opentelemetry support can be disabled by setting the following environment variable: `DD_TRACE_OTEL_ENABLED=False`

### Flask App

```python
import flask
import ddtrace
import opentelemetry

# # The following operations are required when  dd-trace-run is NOT used
# from opentelemetry.trace import set_tracer_provider
# from ddtrace._opentelemetry import TracerProvider
# set_tracer_provider(TracerProvider())
# ddtrace.patch(flask=True)
# import os
# os.environ["OTEL_PYTHON_CONTEXT"] = "ddcontextvars_context"
app = flask.Flask(__name__)
@app.route("/")
def index():
    return "hello"
@app.route("/otel")
def otel():
    oteltracer = opentelemetry.trace.get_tracer(__name__)
    with oteltracer.start_as_current_span("otel-flask-manual-span"):
        with ddtrace.tracer.trace("nested-ddtrace-span"):
            import time
            time.sleep(0.05)
            return "otel", 200
```

#### Sample trace

trace

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

**Comment by mabdinur** [comment]
I can "fix" the failing smoke test by setting `DD_TRACE_OTEL_ENABLED=false`. It appears the `opentelemetry` package in dd-trace-py/tests is conflicting with the `opentelemetry.trace` package defined in the opentelemetry-api

File: ddtrace/opentelemetry/__init__.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from .trace import TracerProvider
```
Comment (mabdinur) [inline ddtrace/opentelemetry/__init__.py:1]:
```suggestion
# TODO: add module level docstring when this tracer provider is GA
from .trace import TracerProvider
```

**Review by emmettbutler — DISMISSED** [review]

File: ddtrace/bootstrap/sitecustomize.py [hunk 2]
```diff
@@ -223,6 +223,13 @@ def update_patched_modules():
         env_tags = os.getenv("DD_TRACE_GLOBAL_TAGS")
         tracer.set_tags(parse_tags_str(env_tags))
 
+    if sys.version_info >= (3, 7) and asbool(os.getenv("DD_TRACE_OTEL_ENABLED", False)):
+        from opentelemetry.trace import set_tracer_provider
+
+        from ddtrace._opentelemetry.trace import TracerProvider
```
Comment (mabdinur) [inline ddtrace/bootstrap/sitecustomize.py:229]: set env var

File: tests/opentelemetry/conftest.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+import opentelemetry
+import pytest
+
+from ddtrace._opentelemetry import TracerProvider
+
+
+TRACER_PROVIDER = TracerProvider()
+# set_tracer_provider can only be called once
+opentelemetry.trace.set_tracer_provider(TRACER_PROVIDER)
```
Comment (mabdinur) [inline tests/opentelemetry/conftest.py:11]: revisit this

File: ddtrace/bootstrap/sitecustomize.py [hunk 4]
```diff
@@ -223,6 +223,13 @@ def update_patched_modules():
         env_tags = os.getenv("DD_TRACE_GLOBAL_TAGS")
         tracer.set_tags(parse_tags_str(env_tags))
 
+    if sys.version_info >= (3, 7) and asbool(os.getenv("DD_TRACE_OTEL_ENABLED", False)):
+        from opentelemetry.trace import set_tracer_provider
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply ddtrace/bootstrap/sitecustomize.py:227]: For `python>=3.7` the opentelemetry api should always be available. The api is lightweight and contains [very few dependencies](https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/pyproject.toml#L27), so I decided to include it as a required dependency in [dd-trace-py/setup.py](https://github.com/DataDog/dd-trace-py/blob/59893d6af7ad0f8d740aa797fa626d6eb7acecf9/setup.py#L373).

File: tests/opentelemetry/conftest.py (hunk 3 above)
Comment (mabdinur) [reply tests/opentelemetry/conftest.py:11]: Using subprocess tests requires a lot of duplication and makes the tests harder to read. I will punt on this for now.

File: setup.py [hunk 5]
```diff
@@ -406,6 +407,12 @@ def get_exts_for(name):
             "ddtrace = ddtrace.contrib.pytest.plugin",
             "ddtrace.pytest_bdd = ddtrace.contrib.pytest_bdd.plugin",
         ],
+        "gevent.plugins.monkey.did_patch_all": [
+            "ddtrace_gevent_check = ddtrace_gevent_check:gevent_patch_all",
+        ],
```
Comment (mabdinur) [inline setup.py:412]:
```suggestion
```

File: tests/snapshots/tests.opentelemetry.test_trace.test_distributed_trace_with_flask_app[with_ddtrace_run].json [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+         "flask.endpoint": "otel",
+         "flask.url_rule": "/otel",
+         "flask.version": "2.1.3",
+         "http.method": "GET",
+         "http.route": "/otel",
+         "http.status_code": "200",
+         "http.url": "http://0.0.0.0:8000/otel",
+         "http.useragent": "python-requests/2.28.1",
+         "language": "python",
+         "runtime-id": "61121a8ca225435796d8141e7eb132e4",
```
Comment (mabdinur) [inline tests/snapshots/tests.opentelemetry.test_trace.test_distributed_trace_with_flask_app[with_ddtrace_run].json:47]:
```suggestion
         "runtime-id": "61121a8ca225435796d8141e7eb132e4",
         "span.kind": "server",
```

**Review by emmettbutler — DISMISSED** [review]

File: .circleci/config.yml [hunk 7]
```diff
@@ -407,6 +407,13 @@ jobs:
       - run_test:
           pattern: 'opentracer'
 
+  opentelemetry:
+    <<: *machine_executor
```
Comment (mabdinur) [inline .circleci/config.yml:411]:
```suggestion
    <<: *machine_executor
     parallelism: 4
```

**Review by emmettbutler — APPROVED** [review]

File: setup.py [hunk 8]
```diff
@@ -383,6 +383,7 @@ def get_exts_for(name):
         "ipaddress; python_version<'3.7'",
         "envier",
         "pep562; python_version<'3.7'",
+        "opentelemetry-api>=1; python_version>='3.7'",
```
Comment (majorgreys) [inline setup.py:386]: We should not assume that the API for setting a context remains the same with the next major version. What happens if the context is not set as expected?

```suggestion
        "opentelemetry-api~=1.0; python_version>='3.7'",
```

Comment (mabdinur) [reply setup.py:386]: We’d like to keep the package version as permissive as possible. If we pin this version then our library will fail to install if a customer attempts to use v2.0. The current approach is consistent with opentracing. If v2.0 introduces changes that are incompatible with our library we’ll ask customers to pin their version and we’ll fix the incompatibilities in a patch release.

Comment (mabdinur) [reply setup.py:386]: Also backwards compatibility is a strict requirement of the otel api. So v2.0 should not introduce breaking changes: https://opentelemetry.io/docs/reference/specification/versioning-and-stability/#design-goals. Hopefully…..

**Review by juanjux — APPROVED** [review]

**Pull request merged**